### PR TITLE
[#11754] Instructor creating course: 'Institute' drop-down is empty if there are no active courses

### LINKED ISSUE #11754: Instructor creating course: 'Institute' drop-down is empty if there are no active courses
v8.14.1, production

A user encountered a problem when creating a course. 

image

It looked like a caching issue (the institute field is missing from the form), but during the troubleshooting, I realized that the 'Institute' dropdown will be empty if the user has no active courses (he had archived courses and  a course in the recycle bin though). The dropdown got populated after I restored the deleted course. I'm waiting to see if that fixed his problem. In the meantime, we can look into if this a bug in the current code.

### PR DESCRIPTION
**Pull request by yengah**
Fixes #11754 

located at: src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.html (line 10)
located at: instructor-home-page.component.ts (line 80)

patch for Issue Instructor creating course: 'Institute' drop-down is empty if there are no active courses TEAMMATES/teammates#11754

**Outline of Patch**

This patch solves the issue by ensuring the field "Institution" always shows a value.

Situation prior to patch:
 * Clicking "Add New Course" from HOME enters COURSE page, and immediately displays the form.  Institution is NULL. This is the Issue (bug).
 
Situation after the patch:
 * Clicking "Add New Course" from HOME enters COURSE page, but does NOT immediately display the form.  Clicking "add new course" button displays the form.  Institution is no longer NULL.  Patch applied.

Ensure that you have:
- [x] Read and understood our PR guideline: https://teammates.github.io/teammates/process.html#step-4-submit-a-pr
  - [x] Added the issue number to the "Fixes" keyword above
  - [x] Titled the PR as specified in the abovementioned document
- [x] Made your changes on a branch other than `master` and `release`
- [x] Gone through all the changes in this PR and ensured that:
  - [x] They addressed one (and only one) issue
  - [x] No unintended changes were made
- [x] Run and passed static analysis: `./gradlew lint` and `npm run lint`
- [ ] Added/updated tests, if changes in functionality were involved
- [ ] Added/updated documentation to public APIs (classes, methods, variables), if applicable

### CODE REVIEW

**Review by jack44321 — APPROVED** [review]

**Comment by yengah** [comment]
> Frontend snapshot tests are failing, please look into the failed tests and either fix the code if the change is unintended, or update the snapshots if they are intended.

@zhaojj2209 Hello, what is the "InstructorHomePageComponent"?

File: src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.html [hunk 1]
```diff
@@ -7,9 +7,10 @@ <h1>Home</h1>
   </div>
 </div>
 
+<!-- altered by Albert Yu  -->
```
Comment (zhaojj2209) [inline src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.html:10]: Please remove this. Your GitHub credentials will be reflected in the commit history if this PR is merged.

File: src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.ts [hunk 2]
```diff
@@ -77,6 +77,10 @@ export class InstructorHomePageComponent extends InstructorSessionModalPageCompo
   isNewUser: boolean = false;
   isCopyLoading: boolean = false;
 
+  // altered by Albert Yu
```
Comment (zhaojj2209) [inline src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.ts:80]: Same here, please remove this credit.

**Comment by zhaojj2209** [comment]
> @zhaojj2209 Hello, what is the "InstructorHomePageComponent"?

I assume you are asking this question because of the frontend test error message and are unsure of where to find the corresponding test:
Screenshot 2022-11-03 at 14 53 30

In which case the answer can be found in the same error message:
Screenshot 2022-11-03 at 14 53 56

`instructor-home-page.component.spec.ts` contains the tests for `instructor-home-page.component.html` and `instructor-home-page.component.ts`, which are the two files you have edited in this PR.

**Review by zhaojj2209 — CHANGES_REQUESTED** [review]

**Comment by yengah** [comment]
> > @zhaojj2209 Hello, what is the "InstructorHomePageComponent"?
> 
> I assume you are asking this question because of the frontend test error message and are unsure of where to find the corresponding test: Screenshot 2022-11-03 at 14 53 30
> 
> In which case the answer can be found in the same error message: Screenshot 2022-11-03 at 14 53 56
> 
> `instructor-home-page.component.spec.ts` contains the tests for `instructor-home-page.component.html` and `instructor-home-page.component.ts`, which are the two files you have edited in this PR.

@zhaojj2209 No, that's not what I meant.  What does "InstructorHomePageComponent" point to?  Which component?

**Comment by zhaojj2209** [comment]
> @zhaojj2209 No, that's not what I meant. What does "InstructorHomePageComponent" point to? Which component?

Still not sure what you're asking about. If you're asking about where to find it in the codebase, `InstructorHomePageComponent` is in `instructor-home-page.component.ts`? If you're asking about where to find it on the UI, it's the UI component that you see when on the instructor home page?

> The code needs some comments..

Any suggestions on what comments to add?

**Comment by yengah** [comment]
> > @zhaojj2209 No, that's not what I meant. What does "InstructorHomePageComponent" point to? Which component?
> 
> Still not sure what you're asking about. If you're asking about where to find it in the codebase, `InstructorHomePageComponent` is in `instructor-home-page.component.ts`? If you're asking about where to find it on the UI, it's the UI component that you see when on the instructor home page?
> 
> > The code needs some comments..
> 
> Any suggestions on what comments to add?

Hello again, this patch will never pass those snapshot tests, unless they are modified.  I hesitate to modify the tests themselves, for reasons you ought to understand.  Please modify the tests if you want this patch to get through Teammates.  Thank you.

**Comment by damithc** [comment]
> I hesitate to modify the tests themselves, for reasons you ought to understand. 

@yengah What are those reasons?

**Comment by yengah** [comment]
> > I hesitate to modify the tests themselves, for reasons you ought to understand.
> 
> @yengah What are those reasons?

These tests really should not change all that often @damithc.  Logically if you think about it this makes good sense.  This is because these automated tests should be fairly foundational, as they are repeated over and over and over again..

**Comment by zhaojj2209** [comment]
> These tests really should not change all that often. Logically if you think about it this makes good sense. This is because these automated tests should be fairly foundational, as they are repeated over and over and over again..

Please read up on snapshot testing. Snapshot tests are used to guard against *unexpected* changes in UI components; when the changes are intended, the snapshot tests should be updated instead (which was what I've already mentioned when first informing you of the failed tests). Regardless of your hesitancy, the onus is on you to update the tests, we won't do it for you.

On another note, while this PR does resolve the issue, it's not ideal UX wise. The "Add New Course" button on the Home page now merely redirects users to the Courses page without opening the form, so users now have to click the "Add New Course" button twice. If possible, a solution where the Home page "Add New Course" button redirects users to the Courses page, opens the course form, and populates the institute dropdown would be preferred.

**Comment by damithc** [comment]
> Please read up on snapshot testing. Snapshot tests are used to guard against _unexpected_ changes in UI components; when the changes are intended, the snapshot tests should be updated instead (which was what I've already mentioned when first informing you of the failed tests). Regardless of your hesitancy, the onus is on you to update the tests, we won't do it for you.

@yengah Yup, aligned to what @zhaojj2209 says here, different tests have different purposes; some tests need to evolve as the application behavior evolves. 
And yes, PR authors are expected to follow the norms of the project. If you want to propose a change to those norms (we are open to them too -- certainly there can be room for improvement in how we do things), they should be posted separately for discussion.

**Comment by yengah** [comment]
> 

@damithc @zhaojj2209 I have a conclusion: because this PR is not logical UX-wise, it would be unwise to change the snapshot test code.

**Comment by damithc** [comment]
> @damithc @zhaojj2209 I have a conclusion: because this PR is not logical UX-wise, it would be unwise to change the snapshot test code.

@yengah Yes. In fact, it would be unwise to merge it at all, in its current state. The remedy:
1. Update the PR to use a better UX, as suggested by @zhaojj2209 .
2. Update the tests accordingly

@zhaojj2209 any further inputs/alternative suggestions?

**Comment by zhaojj2209** [comment]
An ideal solution would be to populate the institute dropdown while still ensuring that the form is shown when clicking the Home page "Add New Course" button.

**Comment by zhaojj2209** [comment]
Closing due to inactivity. Feel free to reopen the PR if you would like to resume working on it.

**Pull request closed without merging**